proposals/lip-0044.md

```diff
@@ -190,234 +216,320 @@ genesisDataSchema = {
+            },
+            topics=[validatorAddress]
+        )
+        raise Exception('Invalid proof of possession for the given BLS key.')
 
     validatorAccount = {
         generatorKey: generatorKey,
         blsKey: blsKey,
     }
-    insert validatorAccount in the validators data substore with storeKey = validatorAddress
-    insert validatorAddress in the registered BLS keys substore with storeKey = blsKey
-
-    return True
+    create an entry in the validators data substore with storeKey = validatorAddress and storeValue = validatorAccount
+    create an entry in the registered BLS keys data substore with storeKey = blsKey and storeValue = validatorAddress
+    emitEvent(
+        moduleID = MODULE_ID_VALIDATORS,
+        typeID = TYPE_ID_GENERATOR_KEY_REGISTRATION,
+        data = {
+            "generatorKey": generatorKey,
+            "result": KEY_REG_RESULT_SUCCESS
+        },
+        topics=[validatorAddress]
+    )
+    emitEvent(
+        moduleID = MODULE_ID_VALIDATORS,
+        typeID = TYPE_ID_BLS_KEY_REGISTRATION,
+        data = {
+            "blsKey": blsKey,
+            "result": KEY_REG_RESULT_SUCCESS
+        },
+        topics=[validatorAddress]
+    )
 ```
 
 The function [`PopVerify`][bls-specs-v4-popverify] is part of the [BLS signature scheme][lip-0038#public-key-registration].
 
 #### registerValidatorWithoutBLSKey
 
 This function creates a new validator account in the validators data substore, where the `blsKey` is set to `INVALID_BLS_KEY`. It is called as part of the [genesis block processing of the DPoS module][lip-0057] for the snapshot block used for the migration from Lisk Core 3 to Lisk Core 4. It should not be called on any other chain nor for any other block.
-
-##### Parameters
-
-This function has the following input parameters in the order given below:
-
-* `validatorAddress`: A 20 bytes value identifying the validator account in the validators data substore.
-* `generatorKey`: A 32 bytes value, specifying the generator key of the validator.
-
-##### Returns
-
-* `success`: A boolean, indicating success or failure of the execution.
-
-##### Execution
-
-This function checks that there is no account already registered for the input `validatorAddress`. Then, it creates a new validator account in the validators data substore, where `blsKey` is set to `INVALID_BLS_KEY`. If this function is called on a chain that is not the mainchain, then this registration fails.
+It checks that there is no account already registered for the input `validatorAddress`. Then, it creates a new validator account in the validators data substore, with `blsKey` set to `INVALID_BLS_KEY` and the generator key set to the input `generatorKey`. If this function is called on a chain that is not the mainchain, then this registration fails.
 
 ```python
-registerValidatorWithoutBLSKey(validatorAddress, generatorKey):
+def registerValidatorWithoutBLSKey(validatorAddress: Address, generatorKey: PublicKeyEd25519) -> None:
```

Comment:
I am now wondering whether it makes sense here to have events at all for this API, in particular, the failing events. This API should only be called during the Lisk Mainnet migration and if this call fails, the snapshot block is invalid. In case of successful block processing, we have to keep in mind that these will be a lot of events emitted by that block.

Reply:
makes sense. I will leave the check and error but remove the event.